working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1302,15 +1238,24 @@ The context type schema for a pattern `p` is:
         and `V` is the greatest lower bound of the context type schemas of all
         value subpatterns.
 
-*   **Record binder or matcher**:
-    *   If the pattern has any positional fields, then the base type schema is
-        `Destructure_n_<F...>` where `_n_` is the number of fields and `F...` is
-        the context type schemas of all of the positional fields.
-    *   Else the base type schema is `Object?`.
-    *   The base type schema is extended with getters for each named field
-        subpattern in `p` where each getter's type schema is the type schema of
-        the corresponding subpattern. (If there is no subpattern because it's
-        an implicit variable pattern like `(field:)`, the type schema is `?`.)
+*   **Record binder or matcher**: A record type scheme with positional and named
+    fields corresponding to the type schemas of the corresponding field
+    subpatterns. (For implicit variable named field subpatterns like `(field:)`,
+    the type schema is `?` for that field.)
+
+    *Note that the type schema will be a record type even when the matched value
+    type isn't a record, as in:*
+
+    ```dart
+    var p = Point(x: 1, y: 2);
+    var (x: x, y: y) = p;
+    ```
+
+    *The record type schema inferred from the pattern may be used to infer
+    record field types on the initialized value if the value is a record
+    literal, and may otherwise be captured. But if the initializer type isn't a
+    record, then the record type context schema inferred from the pattern ends
+    up being ignored.*
```

Comment:
I did have this in the initial proposal (in an extremely half-baked form) and had examples exactly like that in mind.

I took it out because it would mean adding actual structural typing to context types and that felt a little... adventurous for my comfort zone. I'm going to spend time talking to @leafpetersen this week about inference for patterns and we'll talk about this.

Reply:
FYI, @johnniwinther and I had a conversation yesterday about extending context types in some ways that would probably allow this sort of thing to work.  I'd be glad to share his ideas with you if you're interested (or maybe @leafpetersen already knows about it).